[FLINK-32381][runtime] Replaces FatalErrorHandler by Listener.onError

XComp opened the pull request:
## PR Hierarchy

* https://github.com/apache/flink/pull/21742
* https://github.com/apache/flink/pull/22379
* https://github.com/apache/flink/pull/22422
* https://github.com/apache/flink/pull/22380
* https://github.com/apache/flink/pull/22623
* https://github.com/apache/flink/pull/22384
* https://github.com/apache/flink/pull/22390
* https://github.com/apache/flink/pull/22404
* https://github.com/apache/flink/pull/22601
* https://github.com/apache/flink/pull/22642
* https://github.com/apache/flink/pull/22640
* https://github.com/apache/flink/pull/22656
* https://github.com/apache/flink/pull/22828
* https://github.com/apache/flink/pull/22830
* https://github.com/apache/flink/pull/22829
* https://github.com/apache/flink/pull/22661
* https://github.com/apache/flink/pull/22844
* https://github.com/apache/flink/pull/22848
* https://github.com/apache/flink/pull/22873
* === THIS PR === FLINK-32381

## What is the purpose of the change

The driver has two different ways to communicate with the `DefaultLeaderElectionService` right now:
(1) Through the `Listener` interface
(2) Through the `FatalErrorHandler`

This PR merges the two ways into a single communication channel through the `Listener` interface

## Brief change log

* Removed `FatalErrorHandler` field from driver and its constructor
* Adds `onError` to `Listener` interface

## Verifying this change

* `TestingLeaderElectionListener` implements `onError` and some helper methods to retrieve any outstanding error
* `TestingLeaderElectionListener` is used in tests where `FatalErrorHandler` was used beforehand

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

XComp commented:
Force-pushed the rebase to most-recent version of FLINK-32419.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionTest.java:
```diff
@@ -409,20 +408,19 @@ void testExceptionForwarding() throws Exception {
+                assertThat(errorEvent.getError()).isEqualTo(testException);
             } finally {
-                fatalErrorHandler.clearError();
+                electionEventHandler.throwIfErrorEventHappened();
```
This change will cause all the test methods to change there signature from throwing `Throwable` instead of `Exception`. But that's what we want because we also would want to forward Errors, not only Exceptions.

zentol replied:
> But that's what we want because we also would want to forward Errors, not only Exceptions.

Could we just wrap everything into an exception? :thinking:

zentol replied:
Or even better, an unchecked exception?

XComp replied:
I pushed the changes related to your recommendation.

XComp commented:
Force-pushed the rebase to most-recent version of [FLINK-32419](https://issues.apache.org/jira/browse/FLINK-32419).

XComp commented:
Force-pushed rebase to fix the checkstyle issue that was introduced in the previous version of FLINK-32419

XComp commented:
And I had to fix an unused import

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingLeaderElectionListener.java:
```diff
@@ -61,6 +66,19 @@ private void put(LeaderElectionEvent leaderElectionEvent) {
+    public <T> T assertNextEvent(Class<T> expectedEventClass) throws InterruptedException {
+        final LeaderElectionEvent leaderElectionEvent = leaderElectionEvents.take();
+
+        if (expectedEventClass.isAssignableFrom(leaderElectionEvent.getClass())) {
+            return expectedEventClass.cast(leaderElectionEvent);
+        }
+
+        throw new AssertionError(
+                String.format(
+                        "The next event didn't match the expected event type %s.",
+                        expectedEventClass.getSimpleName()));
```
```
assertThat(leaderElectionEvent.getClass).isInstanceOf(expectedEventClass);

return <cast>
```

XComp commented:
Thanks for the approval. I force-pushed a rebase to `master` (with FLINK-32419 being included in `master`) for a final CI run. I addressed your comment accordingly.
